src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -439,13 +439,29 @@ sliceSepBy _cond = undefined
 
 -- | See 'Streamly.Internal.Data.Parser.sliceBeginWith'.
 --
--- /Unimplemented/
+-- /Internal/
 --
 {-# INLINABLE sliceBeginWith #-}
-sliceBeginWith ::
-    -- Monad m =>
-    (a -> Bool) -> Fold m a b -> Parser m a b
-sliceBeginWith = undefined
+sliceBeginWith :: Monad m => (a -> Bool) -> Fold m a b -> Parser m a b
+sliceBeginWith cond (Fold fstep finitial fextract) =
+
+    Parser step initial extract
+
+    where
+
+    initial = Tuple' True <$> finitial
+
+    step (Tuple' isFirstElement s) a =
+        if not (cond a) || isFirstElement
+        then
+            do
+                res <- fstep s a
+                case res of
+                    FL.Partial s' -> return $ Partial 0 (Tuple' False s')
+                    FL.Done b -> return $ Done 0 b
+        else Done 1 <$> fextract s
+
+    extract (Tuple' _ s) = fextract s
```

Comment:
1. You can use multiple datatypes. It's easier to read and understand.
2. Use s1 and s2 instead of s' and s''
3. You can move `return` out of an equation if it is common to all branches.

```suggestion
sliceBeginWith :: Monad m => (a -> Bool) -> Fold m a b -> Parser m a b
sliceBeginWith cond (Fold fstep finitial fextract) = 
    Parser step initial extract

    where

    initial = Left' <$> finitial

    {-# INLINE process #-}
    process s a = do
        res <- fstep s a
        return
            $ case res of
                  FL.Partial s1 -> Partial 0 (Right' s1)
                  FL.Done b -> Done 0 b

    step (Left' s) a = process s a
    step (Right' s) a =
        if not (cond a)
        then process s a
        else Done 1 <$> fextract s

    extract (Left' s) = fextract s
    extract (Right' s) = fextract s
```